File: Collections/ResizingBuffer.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace Collections
{
    public class ResizingBuffer<T> : IQueue<T> where T : class
    {
        // Also present in CircularBuffer
        private T[] buffer;
        private int head = 0;  // read pointer
        private int tail = 0;  // write pointer

        private int size = 0;

        /// <summary>
        /// Initialises an empty resizing buffer.
        /// </summary>
        /// <param name="capacity">Initial capacity with default value</param>
        public ResizingBuffer(int capacity = 10)
        {
            if (capacity < 1)
            {
                throw new ArgumentOutOfRangeException();
            }
            buffer = new T[capacity];
        }

        public ResizingBuffer(params T[] elements) : this(Math.Max(1, elements.Length))
        {
            foreach (var element in elements)
                Enqueue(element);
        }

        private void Resize(int size)
        {
            var temp = new T[size];
            for (int i = 0; i < this.size; i++)
            {
                temp[i] = buffer[(head + i) % buffer.Length];
            }
            buffer = temp;
            head = 0;
            tail = this.size;
        }

        public T Dequeue()
        {
            if (IsEmpty())
            {
                throw new InvalidOperationException("Empty buffer");
            }

            if (size == buffer.Length / 4)
            {
                Resize(buffer.Length / 2);
            }

            size--;
            var returnValue = buffer[head];
            buffer[head] = null;  // prevent loitering
            head = (head + 1) % Capacity();
            return returnValue;
        }

        public void Enqueue(T payload)
        {
            if (IsFull())
            {
                Resize(2 * Capacity());
            }
            size++;
            buffer[tail] = payload;
            tail = (tail + 1) % Capacity();
        }

        public T Peek()
        {
            if (IsEmpty())
            {
                throw new InvalidOperationException("Empty buffer");
            }
            return buffer[head];
        }

        public void Push(T payload)
        {
            Enqueue(payload);
        }

        public T Pop() => Dequeue();

        public int Size() => size;

        public bool IsEmpty() => size == 0;

        private bool IsFull() => size == buffer.Length;  // Not relevant to client

        public int Capacity() => buffer.Length;

        public IEnumerator<T> GetEnumerator()
        {
            var current = head;
            while (current != tail)
            {
                yield return buffer[current];
                current = (current + 1) % Capacity();
            }
        }

        IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
    }
}

File: src/main/java/in/bfdes/collections/ResizingBuffer.java
package in.bfdes.collections;

import java.util.Iterator;
import java.util.NoSuchElementException;

public class ResizingBuffer<T> implements Iterable<T> {
    private T[] buffer;
    private int head = 0;
    private int tail = 0;
    private int size = 0;

    private class BufferIterator implements Iterator<T> {
        private int next = head;

        @Override
        public boolean hasNext() {
            return next != tail;
        }

        @Override
        public T next() {
            if (!hasNext())
                throw new NoSuchElementException();
            T item = buffer[next];
            next = (next + 1) % buffer.length;
            return item;
        }
    }

    public ResizingBuffer(int initialCapacity) {
        if (initialCapacity < 1)
            throw new IllegalArgumentException();
        buffer = (T[]) new Object[initialCapacity];
    }

    public ResizingBuffer() {
        this(10);
    }

    public ResizingBuffer(T[] initialItems) {
        this(Math.max(1, initialItems.length));
        for (var item : initialItems)
            push(item);
    }

    public ResizingBuffer(Iterable<T> initialItems) {
        this();
        for (var item : initialItems)
            push(item);
    }

    public int size() {
        return size;
    }

    public boolean isEmpty() {
        return size == 0;
    }

    private boolean isFull() {
        return size == buffer.length;
    }

    public T peek() {
        if (isEmpty())
            throw new NoSuchElementException();
        return buffer[head];
    }

    public T pop() {
        if (isEmpty())
            throw new NoSuchElementException();
        var item = buffer[head];
        buffer[head] = null;  // prevents loitering
        head = (head + 1) % buffer.length;
        size--;
        if (size == buffer.length / 4)
            resize(buffer.length / 2);
        return item;
    }

    public void push(T item) {
        if (isFull())
            resize(2 * buffer.length);
        buffer[tail] = item;
        tail = (tail + 1) % buffer.length;
        size++;
    }

    public T dequeue() {
        return pop();
    }

    public void enqueue(T item) {
        push(item);
    }

    private void resize(int newCapacity) {
        var newBuffer = (T[]) new Object[newCapacity];
        for (int i = 0; i < size; i++) {
            newBuffer[i] = buffer[(head + i) % buffer.length];
        }
        buffer = newBuffer;
        head = 0;
        tail = size - 1;
    }

    @Override
    public Iterator<T> iterator() {
        return new BufferIterator();
    }
}
